src/os/bluestore/BlueStore.cc

```diff
@@ -16327,7 +16347,7 @@ void BlueStore::_wctx_finish(
     // longer allocated.  Note that this will leave behind edge bits
     // that are no longer referenced but not deallocated (until they
     // age out of the cache naturally).
-    b->discard_unallocated(c.get());
+    // b->discard_unallocated(c.get(), o.get(), lo.e.logical_offset);
```

Comment:
Thanks for pointing it out, this was a memory leak that surprisingly didn’t affect the logic because writes just overwrite these buffers. discard_unallocated is enefficient with newest semantics because it would have to find overlaping segments between and old extent and pexents. I changed it so that truncate and zero do really remove buffers related to removed range since other punch_hole cases like write do just refill punched extent